### PR TITLE
Upgrading to JDK 21 nb-javac

### PR DESCRIPTION
**Pull request by lahodaj**
This is an attempt to upgrade to JDK 21 nb-javac.

---
**^Add meaningful description above**

By opening a pull request you confirm that, unless explicitly stated otherwise, the changes -

 - are all your own work, and you have the right to contribute them.
 - are contributed solely under the terms and conditions of the Apache License 2.0 (see section 5 of the license for more information).

Please make sure (eg. `git log`) that all commits have a valid name and email address for you in the Author field.

If you're a first time contributor, see the Contributing guidelines for more information.

If you're a committer, please label the PR before pressing "Create pull request" so that the right test jobs can run.

### CODE REVIEW

**Comment by mbien** [comment]
@jlahoda @lahodaj if you want to use both accounts, we probably have to figure something out. Should we add collaborator access for your second account? What account should we use to ping you? You can also mail me directly if you want. 
cc @neilcsmith-net

**Comment by neilcsmith-net** [comment]
I think we need to require the account that is linked to ASF for a variety of reasons.

File: java/libs.javacapi/external/binaries-list
Comment (matthiasblaesing) [inline java/libs.javacapi/external/binaries-list:18]: The URL based download should be a fallback, please use our normal maven syntax for artifacts from maven central:

```suggestion
2D8ACD8AAD9C74FDAAA9D7BBAF4278A57E2D86E0 com.dukescript.nbjavac:nb-javac:jdk-21+26:api
9A9BAB9DC8497962967345A59CDD8CBC9D73A40D com.dukescript.nbjavac:nb-javac:jdk-21+26
```

File: java/java.source.base/test/unit/src/org/netbeans/api/java/source/TreeUtilitiesTest.java
Comment (mbien) [inline java/java.source.base/test/unit/src/org/netbeans/api/java/source/TreeUtilitiesTest.java:106]: this will set it to `null` most of the time since `this.sourceLevel` it has no default value. It depends also on what order the tests run, since `testPathForInUnnamedClass` is setting it to `21`, everything before that will use `null` i believe.

(not tested, only looking at it through github)

File: java/java.source.base/test/unit/src/org/netbeans/api/java/source/gen/ClassMemberTest.java
Comment (mbien) [inline java/java.source.base/test/unit/src/org/netbeans/api/java/source/gen/ClassMemberTest.java:1859]: same issue here. Value will depend on test order.

File: java/java.editor.base/test/unit/src/org/netbeans/modules/java/editor/base/semantic/UnusedDetectorTest.java
Comment (mbien) [inline java/java.editor.base/test/unit/src/org/netbeans/modules/java/editor/base/semantic/UnusedDetectorTest.java:533]: similar here, tests will use `1.8`, until this method runs.

File: java/java.source.base/test/unit/src/org/netbeans/api/java/source/TreeUtilitiesTest.java
Comment (lahodaj) [reply java/java.source.base/test/unit/src/org/netbeans/api/java/source/TreeUtilitiesTest.java:106]: The intent here is not to change anything for existing tests (as we can't be sure the tests would still be testing whatever it was supposed to test). I can wrap the call to `setSourceLevel` with a `if (sourceLevel != null)`, which should make the intent more clear.

Please note that every test method gets its own instance of the class, and hence will start with the `sourceLevel` field initialized to `null`. The test order should not have any effect on the results, I believe.

File: java/java.source.base/test/unit/src/org/netbeans/api/java/source/gen/ClassMemberTest.java
Comment (lahodaj) [reply java/java.source.base/test/unit/src/org/netbeans/api/java/source/gen/ClassMemberTest.java:1859]: I don't think there should be any dependence on test order. Every test gets its own instanceof of the class (and its own working directory), I believe, and hence every test starts with `sourceLevel == null`, which is (was) the default so far. If the test sets a source level value, it should be used only for that test, and no other.

File: java/java.source.base/test/unit/src/org/netbeans/api/java/source/TreeUtilitiesTest.java
Comment (mbien) [reply java/java.source.base/test/unit/src/org/netbeans/api/java/source/TreeUtilitiesTest.java:106]: > Please note that every test method gets its own instance of the class,

ok thanks - my mistake. I usually have parameter setup in the `@Before` method (or equivalent) to make it explicit what the non-constant parameters are. But if the whole class resets for every tests case in NB's test setup its essentially doing the same. All good.

**Comment by mbien** [comment]
before merge, please don't forget to bump the version at
https://github.com/apache/netbeans/blob/8df65b70ed0d17f0e641b961b0ede8ef89ba640d/java/java.source.base/src/org/netbeans/modules/java/source/NoJavacHelper.java#L37

and module description as in https://github.com/apache/netbeans/pull/4010/commits/2f1885d65e56067b0c2c85698d3b6c0920baa338

File: java/java.completion/src/org/netbeans/modules/java/completion/JavaCompletionTask.java
Comment (dbalek) [inline java/java.completion/src/org/netbeans/modules/java/completion/JavaCompletionTask.java:1]: `JavaCompletionTask` will definitely need more changes to work correctly with 21. I will prepare a subsequent PR once this PR gets merged.

**Review by dbalek — APPROVED** [review]

**Comment by neilcsmith-net** [comment]
Adding do not merge label to stop this being merged accidentally.  This needs squashing, committing (and possibly a new PR) under the account attached to the ASF membership.

**Comment by lahodaj** [comment]
Not completely clear why a PR must be linked to an ASF membership, but it is the case now.

**Comment by neilcsmith-net** [comment]
@lahodaj because I'm not sure we have any other way to authenticate and log the PR and commit email against your ICLA and PMC status?!

**Comment by mbien** [comment]
> Not completely clear why a PR must be linked to an ASF membership, but it is the case now.

@lahodaj one account had commit rights, the other didn't. This includes things like running CI, reviewing PRs, labeling, linking issues, it decides if you show up in the reviewer list for others (...).

From the project side it is also the only way to check ICLA status as @neilcsmith-net mentioned. Which is going to be useful as soon jlahodaj2_deluxe shows up.

Thanks for adding your second gh handle to the apache account - this solves this problem I believe.

**Comment by arvindaprameya** [comment]
Thank you for making this change. Approved

**Comment by lahodaj** [comment]
@neilcsmith-net - are there objections against this PR? Thanks!

**Comment by neilcsmith-net** [comment]
@lahodaj no objections from me, just wanted the authentication of the account resolved.  Removed the label.

Would prefer we didn't touch .sig files during release phases, but I can handle the potential conflict later.

**Comment by matthiasblaesing** [comment]
I think I see one fallout of the update:

```
SEVERE [org.openide.util.Exceptions]
java.lang.ClassCastException: class com.sun.tools.javac.util.Names cannot be cast to class org.netbeans.lib.nbjavac.services.NBNames (com.sun.tools.javac.util.Names is in unnamed module of loader org.netbeans.StandardModule$OneModuleClassLoader @597da675; org.netbeans.lib.nbjavac.services.NBNames is in unnamed module of loader org.netbeans.StandardModule$OneModuleClassLoader @2c287d54)
	at org.netbeans.api.java.source.SourceUtils.isMainMethod(SourceUtils.java:1033)
	at org.netbeans.modules.java.source.usages.SourceAnalyzerFactory$UsagesVisitor.visitMethod(SourceAnalyzerFactory.java:772)
	at org.netbeans.modules.java.source.usages.SourceAnalyzerFactory$UsagesVisitor.visitMethod(SourceAnalyzerFactory.java:276)
	at com.sun.tools.javac.tree.JCTree$JCMethodDecl.accept(JCTree.java:948)
	at com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:92)
	at org.netbeans.modules.java.source.usages.SourceAnalyzerFactory$UsagesVisitor.scan(SourceAnalyzerFactory.java:389)
	at org.netbeans.modules.java.source.usages.SourceAnalyzerFactory$UsagesVisitor.scan(SourceAnalyzerFactory.java:276)
	at com.sun.source.util.TreeScanner.scanAndReduce(TreeScanner.java:96)
	at com.sun.source.util.TreeScanner.scan(TreeScanner.java:111)
	at org.netbeans.modules.java.source.usages.SourceAnalyzerFactory$UsagesVisitor.visitClass(SourceAnalyzerFactory.java:721)
	at org.netbeans.modules.java.source.usages.SourceAnalyzerFactory$UsagesVisitor.visitClass(SourceAnalyzerFactory.java:276)
```

This was reported when I had a class with a `main` method open.

**Comment by mbien** [comment]
@matthiasblaesing I can't reproduce this. I changed the code slightly so that I can set a breakpoint there and the type seems to be always `NBNames`:
```java
        Names names = ((Symbol.MethodSymbol)method).name.table.names;
        Context ctx = ((NBNames) names).getContext();
```
test:
 - create new class
 - add main method

**Comment by matthiasblaesing** [comment]
@mbien thanks for checking. I'm currently running NetBeans master at work and at home and can't reproduce anymore. Sorry for the noise.

**Comment by mbien** [comment]
@matthiasblaesing all good! thanks for verifying.

**Comment by ebarboni** [comment]
@jlahoda sorry for being late in this already merged PR but the apidoc generation is broken 

javadoc cannot use custom bootstrap cp (If I remember well), so in the following file Convert.Validation is not found, (using jdk21 as javadoc tool)
java/java.source.base/src/org/netbeans/modules/java/source/usages/ClassFileUtil.java:24
 import com.sun.tools.javac.util.Convert.Validation;

**Pull request merged**
